sharktank/sharktank/layers/paged_attention.py

```diff
@@ -367,64 +227,620 @@ def write_timestep(
             else:
                 page_table.index_put_(indices=(index,), values=values)
 
-        return
+
+class ShardedCache:
+    def __init__(
+        self,
+        *,
+        shard_count: int,
+        transformer_block_count: int,
+        attn_head_count: int,
+        attn_head_dim: int,
+        devices: List[int] | None = None,
+        cache_partition_count: int = 2,
+        block_seq_stride: int = 16,
+        cache_dtype: torch.dtype = torch.float32,
+        device: Optional[torch.device] = None,
+    ):
+        caches = []
+        for i in range(shard_count):
+            start = i * attn_head_count // shard_count
+            end = (i + 1) * attn_head_count // shard_count
+            sharded_attn_head_count = end - start
+
+            cache = KVCache(
```

Comment:
This should be getting passed devices[i], otherwise placement won't be correct.

Reply:
No, `SplitPrimitiveTensor` does not rely on the `ReplicatedTensor` device placement and afixes the device placement information itself. It purely requires that it is consist with its own tracking / device placement of the tensors.